pkg/envoy/xds/ack.go

```diff
@@ -139,15 +139,17 @@ func (m *AckingResourceMutatorWrapper) Upsert(typeURL string, resourceName strin
 	m.locker.Lock()
 	defer m.locker.Unlock()
 
+	version, _, revert := m.mutator.Upsert(typeURL, resourceName, resource, true)
+
 	if c == nil {
-		log.WithFields(logrus.Fields{
-			logfields.XDSTypeURL:      typeURL,
-			logfields.XDSResourceName: resourceName,
-		}).Fatal("no completion given to upsert xDS resource")
+		// No completion in upsert, revert also without completion
+		return func(_ *completion.Completion) {
```

Comment:
I think it would be simpler and more consistent to just use with the revert func below. Even if this mutation didn't have a completion, it doesn't mean that you should ignore any completion for reverting it.
Just remove this whole `if` block here, and enclose the stuff below into an `if c != nil`:
```
if c != nil {
    if _, found := m.pendingCompletions[c]; found {
        ...
    }
    ...
    m.pendingCompletions[c] = comp
}
```

Reply:
Right, will make the original revert funcs deal with nil completions as well.

Reply:
Cool, I didn't catch that either.